### ISSUE TITLE
VSC R Debugger not initiating on WINDOWS 10

### ISSUE THREAD

**Issue by gwd999** [ManuelHentschel/VSCode-R-Debugger#131]
I am exhibiting various issues on WINDOWS 10 PRO 20H2 
VS Code: 1.54.2
R version: R-4.0.4 
vscDebugger version:  0.4.5
svSocket branch "for vscDebugger" (installed from your repo fork) vers: 1.0.1    

DEBUG CONSOLE output like the following during "Debug R Function" (same for "Debug R file")
```
"rStrings": {
    "prompt": "<#v\\s\\c>",
    "continue": "<##v\\s\\c>",
    "startup": "<v\\s\\c\\R\\STARTUP>",
    "libraryNotFound": "<v\\s\\c\\LIBRARY\\NOT\\FOUND>",
    "packageName": "vscDebugger"
  },   
```
somehow does not "feel" right, but tbh I am not yet familiar enough with yr code to judge ? 
Should the "libraryNotFound" worry me? 
 
None of the  Debug Configurations (File, Function) will fire up correctly it seems. 
 
I do have a custom R package library .libPath set in Rprofile.site pointing to another partition "D:\"  
I also tried installing "vscDebugger" in the default R_HOME\library - but that did also not fix the issue. 

Debug Config "Attach to R process" throws: 
![image](https://user-images.githubusercontent.com/306826/111077225-0b356c00-84f0-11eb-9b37-8508efe799de.png)

_Originally posted by @gwd999 in https://github.com/ManuelHentschel/VSCode-R-Debugger/issues/124#issuecomment-798942394_

**Comment by ManuelHentschel** [comment]
>  Should the "libraryNotFound" worry me?

Not at all, those strings are just internal information (for debugging the debug extension itself). I should probably remove these, since they can be a bit misleading.


*"Launch R-Workspace"* is the most basic of the default debug configs, so I'd recommend trying to get that one to work first. What's the behavior if you try launching the debugger with this config?

**Comment by gwd999** [comment]
Ouput of DEBUG CONSOLE when clicking "Launch R Workspace"
```
Starting R session...
R Startup:
{
  "path": "C:\\R\\R-4.0.4\\bin\\x64\\R.exe",
  "args": [
    "--quiet",
    "--no-save",
    "--ess"
  ],
  "cwd": "c:\\Users\\gwd\\Documents\\dev\\R\\misc"
}
Initialize Arguments:
{
  "clientID": "vscode",
  "clientName": "Visual Studio Code",
  "adapterID": "R-Debugger",
  "pathFormat": "path",
  "linesStartAt1": true,
  "columnsStartAt1": true,
  "supportsVariableType": true,
  "supportsVariablePaging": true,
  "supportsRunInTerminalRequest": true,
  "locale": "en-us",
  "supportsProgressReporting": true,
  "supportsInvalidatedEvent": true,
  "threadId": 1,
  "extensionVersion": "0.4.5",
  "rStrings": {
    "prompt": "<#v\\s\\c>",
    "continue": "<##v\\s\\c>",
    "startup": "<v\\s\\c\\R\\STARTUP>",
    "libraryNotFound": "<v\\s\\c\\LIBRARY\\NOT\\FOUND>",
    "packageName": "vscDebugger"
  },
  "useJsonSocket": true,
  "jsonHost": "localhost",
  "jsonPort": 55511,
  "useSinkSocket": true,
  "sinkHost": "localhost",
  "sinkPort": 55512
}
Launch Arguments:
{
  "type": "R-Debugger",
  "name": "Launch R-Workspace",
  "request": "launch",
  "debugMode": "workspace",
  "workingDirectory": "C:\\Users\\gwd\\Documents\\dev\\R\\misc",
  "__configurationTarget": 5,
  "allowGlobalDebugging": true,
  "supportsStdoutReading": true,
  "supportsWriteToStdinEvent": true,
  "supportsShowingPromptRequest": true,
  "overwriteHelp": true,
  "__sessionId": "204ea383-6ca4-41f5-aaab-5574d105929c"
}
```

**Comment by ManuelHentschel** [comment]
You could try copying the following settings into `./.vscode/settings.json`:
``` json
{
    "r.debugger.printStdout": "all",
    "r.debugger.printStderr": "all",
    "r.debugger.printSinkSocket": "all",
    "r.debugger.printStdin": "all",
}
```
And checking the output of the debug console again (hopefully a bit more is printed then).

**Comment by ManuelHentschel** [comment]
You could also try launching R in a normal terminal and running e.g. 
```
> x <- 100
> vscDebugger::.vsc.listenForDAP()
Listening on localhost:18721 # hopefully this is the resulting output
```
and then launching the debug mode `Attach to R process`.

**Comment by ManuelHentschel** [comment]
The debug config `Launch R Workspace` is not supposed to run the code in your file, but still show a call stack frame labelled `Global Workspace`.

The debug config `Debug R File` is supposed to run your file.

The debug mode `Debug R Function` only works if you have a `main()` function in your file, e.g.
``` r
main <- function(){
    print("Hello World")
}
```

**Comment by gwd999** [comment]
> You could also try launching R in a normal terminal and running e.g.
> 
> ```
> > x <- 100
> > vscDebugger::.vsc.listenForDAP()
> Listening on localhost:18721 # <- hopefully this is the resulting output
> ```
> 
> and then launching the debug mode `Attach to R process`.
Launching  `Attach to R process` just kills the R process in the Terminal.

**Comment by ManuelHentschel** [comment]
If none of this makes a difference, you can also try installing the latest version of the vscode extension and R package from [here](https://github.com/ManuelHentschel/VSCode-R-Debugger/releases/tag/latest) and [here](https://github.com/ManuelHentschel/vscDebugger/releases/tag/latest), in case your issue was "accidently" fixed by some other change in the meantime.

**Comment by ManuelHentschel** [comment]
> Launching `Attach to R process` just kills the R process in that Terminal which did show `Listening on localhost:18721`.

That's strange, no idea what's going on here, but that's definitely not supposed to happen -.-

**Comment by gwd999** [comment]
> You could try copying the following settings into `./.vscode/settings.json`:
> 
> ```json
> {
>     "r.debugger.printStdout": "all",
>     "r.debugger.printStderr": "all",
>     "r.debugger.printSinkSocket": "all",
>     "r.debugger.printStdin": "all",
> }
> ```
> 
> And checking the output of the debug console again (hopefully a bit more is printed then).

Adding the above resulted in  
![image](https://user-images.githubusercontent.com/306826/111080353-653d2e00-84fe-11eb-85df-f46e8765a7bc.png)

**Comment by gwd999** [comment]
> > Launching `Attach to R process` just kills the R process in that Terminal which did show `Listening on localhost:18721`.
> 
> That's strange, no idea what's going on here, but that's definitely not supposed to happen -.-

The above message popup with "ECONNREFUSED" in it did not look too promising from the start ;-)

**Comment by gwd999** [comment]
> If none of this makes a difference, you can also try installing the latest version of the vscode extension and R package from [here](https://github.com/ManuelHentschel/VSCode-R-Debugger/releases/tag/latest) and [here](https://github.com/ManuelHentschel/vscDebugger/releases/tag/latest), in case your issue was "accidently" fixed by some other change in the meantime.

Installing from these two links did not change the outcome ... :-(

**Comment by ManuelHentschel** [comment]
> 
> 
> > If none of this makes a difference, you can also try installing the latest version of the vscode extension and R package from [here](https://github.com/ManuelHentschel/VSCode-R-Debugger/releases/tag/latest) and [here](https://github.com/ManuelHentschel/vscDebugger/releases/tag/latest), in case your issue was "accidently" fixed by some other change in the meantime.
> 
> Installing from these two links did not change the outcome ... :-(
> Could some other extensions be interfering? Although I hardly can imagine another extension killing that listening on localhost:18721 R terminal process.

If any, the session watcher from `vscode-R` extension might interfere, though I'm using that too.

**Comment by ManuelHentschel** [comment]
> > You could try copying the following settings into `./.vscode/settings.json`:
> > ```json
> > {
> >     "r.debugger.printStdout": "all",
> >     "r.debugger.printStderr": "all",
> >     "r.debugger.printSinkSocket": "all",
> >     "r.debugger.printStdin": "all",
> > }
> > ```
> > 
> > 
> > And checking the output of the debug console again (hopefully a bit more is printed then).
> 
> Adding the above resulted in

Does the output to the debug console change, when you add these settings and run `Launch R-Workspace`?

**Comment by ManuelHentschel** [comment]
This might also show some more info when copied to `.vscode/settings.json`:
``` json
{
    "r.debugger.printStdout": "all",
    "r.debugger.printStderr": "all",
    "r.debugger.printSinkSocket": "all",
    "r.debugger.printStdin": "all",
    "r.debugger.printJsonSocket": "all",
    "r.debugger.logLevelSession": "debug",
    "r.debugger.logLevelRuntime": "debug",
    "r.debugger.logLevelRSession": "debug"
}
```

**Comment by gwd999** [comment]
Output in DEBUG CONSOLE stays the same ... for "Launch R-Workspace"

```
R Startup:
{
  "path": "C:\\R\\R-4.0.4\\bin\\x64\\R.exe",
  "args": [
    "--quiet",
    "--no-save",
    "--ess"
  ],
  "cwd": "c:\\Users\\gwd\\Documents\\dev\\R\\misc"
}
```
Only after clicking "Restart" it gets a little longer and the Call Stack shows up with  second session 

![image](https://user-images.githubusercontent.com/306826/111083975-a6d6d480-8510-11eb-9610-ea8e5f506454.png)




```
Starting R session...
R Startup:
{
  "path": "C:\\R\\R-4.0.4\\bin\\x64\\R.exe",
  "args": [
    "--quiet",
    "--no-save",
    "--ess"
  ],
  "cwd": "c:\\Users\\gwd\\Documents\\dev\\R\\misc"
}
Initialize Arguments:
{
  "clientID": "vscode",
  "clientName": "Visual Studio Code",
  "adapterID": "R-Debugger",
  "pathFormat": "path",
  "linesStartAt1": true,
  "columnsStartAt1": true,
  "supportsVariableType": true,
  "supportsVariablePaging": true,
  "supportsRunInTerminalRequest": true,
  "locale": "en-us",
  "supportsProgressReporting": true,
  "supportsInvalidatedEvent": true,
  "threadId": 1,
  "extensionVersion": "0.4.5",
  "rStrings": {
    "prompt": "<#v\\s\\c>",
    "continue": "<##v\\s\\c>",
    "startup": "<v\\s\\c\\R\\STARTUP>",
    "libraryNotFound": "<v\\s\\c\\LIBRARY\\NOT\\FOUND>",
    "packageName": "vscDebugger"
  },
  "useJsonSocket": true,
  "jsonHost": "localhost",
  "jsonPort": 57422,
  "useSinkSocket": true,
  "sinkHost": "localhost",
  "sinkPort": 57421
}
```

**Comment by gwd999** [comment]
> If any, the session watcher from `vscode-R` extension might interfere, though I'm using that too.

I turned R extension Session Watcher off -> no difference

**Comment by ManuelHentschel** [comment]
I'm a bit out of ideas, since everything looks normal in those logs and no error messages pop up, but something obviously isn't working.

If you are interested, you could try running the extension from code (clone the vscode-R-debugger repo, install npm, run `npm install` and press F5) to see what's going wrong.

Alternatively, I'm currently working on some refactoring that includes better logging, which might help to find the cause of these issues.

**Comment by gwd999** [comment]
Ok I think I found the direction - I had the same issues when testing "remotely" ie in VSC on WSL2 (Ubuntu)
UNTIL I commented out everything in my ".Rprofile" there! 
So I will start investigating - ie stepwise removing the comments there - to see what part of my .Rprofile interferes w your extension. 
Unluckily I had this idea for my WINDOWS setup as well - but there it did not help. 
But maybe if I get a grip on what mixes with it on my Ubuntu setup I can then deduce and further dig into the Windows problem. 
Thanks a lot for your patience and assistance.

**Comment by ManuelHentschel** [comment]
Thanks for the updates, I'm definitely interested! Ideally it might be possible to make the debugger more robust to cope with these packages/configs.

**Comment by ManuelHentschel** [comment]
To avoid problems like this you can use e.g.
``` r
if(Sys.getenv('VSCODE_DEBUG_SESSION')=='1'){
    ### Debugger-specific code here
} else {
    ### Debugger-breaking code here
}
```
in your .Rprofile (this env variable gets set whenever the debugger is responsible for launching R).

**Comment by gwd999** [comment]
> Thanks for the updates, I'm definitely interested! Ideally it might be possible to make the debugger more robust to cope with these packages/configs.

So now that we found the "bad guy" what do you think? Is there some method I could do to keep the best of both - your great extension and that awesome prompt?

**Comment by gwd999** [comment]
Final Remark the only thing left now is when trying to use "Attach to R process" DEBUG CONFIG I still get 
![image](https://user-images.githubusercontent.com/306826/111088467-a0078c00-8527-11eb-804c-47912431ab69.png)
on BOTH (Windows and WSL2 Ubuntu) 
But I can definitely live with that as long as debugging with the other settings is available.

**Comment by ManuelHentschel** [comment]
Thanks for the positive feedback! I think I'll write up a "Troubleshooting" Vignette to include information like this, since the normal readme is rather long already.

**Comment by ManuelHentschel** [comment]
> 
> 
> Final Remark the only thing left now is when trying to use "Attach to R process" DEBUG CONFIG I still get
> ![image](https://user-images.githubusercontent.com/306826/111088467-a0078c00-8527-11eb-804c-47912431ab69.png)
> on BOTH (Windows and WSL2 Ubuntu, no matter what the .Rprofile)
> But I can definitely live with that as long as debugging with the other settings is available.

For this debug mode to work you need to call `.vsc.startWebsocket()` or `.vsc.listenForDAP()` in an existing R session (the debugger will try to attach to that session). If you keep having problems with this that you want to work out, please report in another issue to keep things somewhat organized

**Comment by gwd999** [comment]
> For this debug mode to work you need to call `.vsc.startWebsocket()` or `.vsc.listenForDAP()` in an existing R session (the debugger will try to attach to that session)

Thanks for the hint - but still the same behaviour on this - that "outside" R terminal when I click "attach to R process" will while using  `.vsc.listenForDAP()` crash the R console and when using `.vsc.startWebsocket()` just stops the listening and go back to the input prompt. 
But as I mentioned that is a first world problem imo ...
